geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/ScalaSimpleFeature.scala

```diff
@@ -116,142 +139,180 @@ object ScalaSimpleFeature {
+     * @param reader lazily read attributes, must be thread-safe and already of the appropriate types
+     * @param userDataReader lazily read user data, must be thread-safe
+     *
+     */
+    def this(sft: SimpleFeatureType, id: String, reader: LazyAttributeReader, userDataReader: LazyUserDataReader) = {
+      this(sft)
+      this.id = id
+      this.reader = reader
+      this.userDataReader = userDataReader
     }
 
     override def setAttributeNoConvert(index: Int, value: AnyRef): Unit = {
       if (reader != null) {
         val word = bits.word(index)
         word.synchronized {
           if (word.add(index)) {
-            phaser.arriveAndDeregister()
+            bits.synchronized {
+              if (attributes == null) {
+                attributes = Array.ofDim[AnyRef](sft.getAttributeCount)
+              }
+            }
+            if (count.decrementAndGet() == 0) {
+              // once all attributes have been read, dereference any backing resources so that they can be gc'd
+              reader = null
+              count = null
+            }
           }
         }
       }
       attributes(index) = value
     }
 
+    override def calculateSizeOf(): Long =
+      Sizable.sizeOf(this) + Sizable.deepSizeOf(id, bits, attributes, userData, count, reader, userDataReader)
+  }
+
+  /**
+   * Thread-safe lazy evaluation of attributes
+   */
+  trait LazySimpleFeature extends AbstractSimpleFeature {
+
+    // we synchronize on the low-level words, in order to minimize contention
+    protected val bits: WordBitSet = WordBitSet(getFeatureType.getAttributeCount + 1)
+    protected var attributes: Array[AnyRef] = _
+    protected var userData: java.util.Map[AnyRef, AnyRef] = _
+
+    protected var count = new AtomicInteger(getFeatureType.getAttributeCount)
+    protected var reader: LazyAttributeReader = _
+    protected var userDataReader: LazyUserDataReader = _
+
     override def getAttribute(index: Int): AnyRef = {
       if (reader != null) {
         val word = bits.word(index)
         word.synchronized {
           if (word.add(index)) {
+            bits.synchronized {
+              if (attributes == null) {
+                attributes = Array.ofDim[AnyRef](getFeatureType.getAttributeCount)
+              }
+            }
             attributes(index) = reader.read(index)
-            phaser.arriveAndDeregister()
+            if (count.decrementAndGet() == 0) {
+              // once all attributes have been read, dereference any backing resources so that they can be gc'd
+              reader = null
+              count = null
+            }
           }
         }
       }
       attributes(index)
     }
-  }
 
+    override def getUserData: java.util.Map[AnyRef, AnyRef] = {
+      if (userDataReader != null) {
+        synchronized {
+          if (userData == null) {
+            userData = userDataReader.read()
+            userDataReader == null // dereference any backing resources
+          }
```

Comment:
Do we need to do any handling in the userDataReader != null && userData != null case?  

Can someone call 'setUserData' before the userDataReader is used (and hence released)?

Reply:
there is no `setUserData` method, so we don't have to worry about that...
i think the synchronized block should ensure that if the user data is null, then the userDataReader will not be null, since it is only set to null inside that block. it's possible that multiple threads could hit the unsynchronized `if (userDataReader != null)` check, but i don't _think_ that would matter since they'd block on the synchronized block, and userDataReader can't be null until userData has been read and is thus valid to return.
any subtle threading issues i'm missing here?

Reply:
"Maybe".

This seems like a backwards version of the 'lazy initialization' problem.  For that double-locking is either answer or an anti-pattern.

https://en.wikipedia.org/wiki/Double-checked_locking
https://www.cs.umd.edu/~pugh/java/memoryModel/DoubleCheckedLocking.html

Reply:
In short, I don't know.  I think two threads could both get to the synchronized block at the same time.  The second one should throw an error/NPE (line 291), right?

Reply:
this isn't quite the same as double checked locking, since the second check is a different variable. I *think* this is ok... taking it line-by-line:
```
if (userDataReader != null) {
  synchronized {
    if (userData == null) {
      userData = userDataReader.read()
      userDataReader == null // dereference any backing resources
    }
  }
}
userData
```

1. It's guaranteed that by the userData will be non-null by the time userDataReader is null. Thus any thread that comes through and sees userDataReader as null will return the non-null userData. Potentially there needs to be a volatile keyword on userDataReader? I assumed setting something null would be a single operation, but maybe nt.
2. If userDataReader is not null, then it will wait on a synchronized block. Once it gains the block, it will check for userData being null and read it if needed. If another thread gained the synchronized block first, then it will just continue on.

Reply:
Ah, ok.  My coffee really isn't working today (or this is slightly subtle).  

I have reached your same level of understanding.  I think this is correct.  Also, we are talking about a somewhat crazy case where multiple threads are reading userData at the same time... 

So yeah, we'd have to be 'wrong' and then we'd have to be 'unlucky'. ;)